File: qwopper-master/src/com/slowfrog/qwop/GA.java
package com.slowfrog.qwop;

import java.util.*;
import java.lang.*;

public class GA {
	public static Random random = new Random();
	//initiate the first generation
	public static ArrayList initiate(int populationSize) {
		ArrayList population = new ArrayList();
		
		for (int i=0; i<populationSize; i++) {
			String str = "";
			for (int j=0;j<8;j++) {
				int aNum = random.nextInt(255);
				String theNum = aNum.toBinaryString(aNum);
				str.concat(theNum);
			}
			population.add(str);
		}
		return population;
	}
	
	public static ArrayList fitness(ArrayList population, ArrayList data) {
		ArrayList result = new ArrayList();
		
		for (i=0;i<population.size();i++) {
			String str = data.get(i);
			String[] lst = str.split("\\|");
			result.add(lst[3]);
		}
		return result;
	}
	
	public static ArrayList crossover(ArrayList population, int crossoverRate) {
		ArrayList newPopulation = new ArrayList();
		
		for (int i=0; i<population.size()/2; i++) {
			String a = population.remove(int random.nextInt(population.size()));
			String b = population.remove(int random.nextInt(population.size()));
			if (float random.nextFloat() < crossoverRate) {
				int start = random.nextInt(64);
				int end = start + random.nextInt(64-start);
				String sub = a.substring(start,end);
				a = a.substring(start)+b.substring(start,end)+a.substring(end,64);
				b = b.substring(start)+sub+b.substring(end,64);
			}
			newPopulation.add(a);
			newPopulation.add(b);
		}
		return newPopulation;
	}
	
	public static ArrayList mutation(ArrayList population, int mutationRate) {
		for (int i=0;i<population.size();i++) {
			for (int index=0;index<population[i].length();index++) {
				if (float random.nextFloat() < mutationRate) {
					String str;
					if (population.get(i).charAt(index) == '0') {
						str = population.get(i).substring(index)+'1'+population.get(i).substring(index+1,population.get(i).length());
					} else {
						str = population.get(i).substring(index)+'0'+population.get(i).substring(index+1,population.get(i).length());
					}
					population.remove(i);
					population.add(i,str);
				}
			}
		}
		return population;
	}
	
	public static ArrayList ga(int populationSize, int crossoverRate, int mutationRate) {
		this.populationSize = populationSize;
		this.crossoverRate = crossoverRate;
		this.mutationRate = mutationRate;
		//use aScore to determine the top chromosomes
		this.aScore = 30;
		//determine when consider finding the solution
		this.endPoint = 100;
	}
	
	
}

File: qwopper-master/src/com/slowfrog/qwop/ga.py
import random

def initiate(popSize):
    population = []
    for i in range(popSize):
        population.append( ##represent second of pressing a button, and second of releasing it, in the order of QWOP
                           '{0:08b}'.format(random.randint(0,255)) + '{0:08b}'.format(random.randint(0,255)) +
                           '{0:08b}'.format(random.randint(0,255)) + '{0:08b}'.format(random.randint(0,255)) +
                           '{0:08b}'.format(random.randint(0,255)) + '{0:08b}'.format(random.randint(0,255)) +
                           '{0:08b}'.format(random.randint(0,255)) + '{0:08b}'.format(random.randint(0,255)))
##                           ##represent starting either pressing or releasing a button, in the order of QWOP, 1 is press
##                           [random.randint(0,1),random.randint(0,1),random.randint(0,1),random.randint(0,1)]
##                           ])
    return population

##number represent how far it can run
def fitness(pop, data):
    result = []
    for i in pop:
##placeholder of fitness function
        result.append(random.randint(0,100))
    return result
        

def crossover(population, crossoverRate):
    newPopulation = []
    for i in range(int(len(population)/2)):
        a = population.pop(random.randint(0,len(population)-1))
        b = population.pop(random.randint(0,len(population)-1))
        if random.random() < crossoverRate:
            start = random.randint(0,91)
            end = random.randint(start,91)
            new = a[start:end]
            a = a[:start] + b[start:end] + a[end:]
            b = b[:start] + new + a[end:]
        newPopulation.append(a)
        newPopulation.append(b)
    return newPopulation

##mutation now mutates all digits
def mutation(population, mutationRate):
    for i in range(len(population)):
        for index in range(len(population[i])):
            if random.random() < mutationRate:
                if population[i][index] == '0':
                    population[i] = population[i][:index] + '1' + population[i][index+1:]
                else:
                    population[i] = population[i][:index] + '0' + population[i][index+1:]
    return population

def main():
    populationSize = int(input("Population Size is => "))
    crossoverRate = float(input("Crossover Rate is => "))
    mutationRate = float(input("Mutation Rate is => "))
##use aScore to determine the top chromosomes
    aScore = 30
##determine when consider finding the solution
    endPoint = 99

    
##initialize population
    population = initiate(populationSize)
    data = []
##save the top chromosomes in the first generation
    saved0 = []
    result = fitness(population, data)
    for i in range(len(result)):
        if result[i] > aScore:
            saved0.append(population[i])
    result_copy = result[:]
    result_copy.sort()
    result_copy.reverse()

##if find a solution (this happens rarely from generation1)
    if result_copy[0] > endPoint:
        print("Hey, I found it")
        best_index = result.index(result_copy[0])
        print("the gene is:")
        print(population[best_index])
        print("It can run " + str(result_copy[0]) + "meters.")
        pop = []
        pop.append(population[best_index])
        fitness(pop,data)
        return pop
    
    end = "no"
    while end == "no":
        print("a generation is passed")
##mutation and crossover
        population = mutation(population, mutationRate)    
        population = crossover(population, crossoverRate)
##save the top chromosomes in the first generation
        saved1 = []
        result = fitness(population, data)
        result_copy = result[:]
        result_copy.sort()
        result_copy.reverse()
##save the top 10% if the top 10% is better than aScore
        if result_copy[int(len(result_copy)/10)] > aScore:
            for i in range(len(result)):
                if result[i] > result_copy[int(len(result_copy)/10)]:
                    saved1.append(population[i])
##save the ones that is better than aScore if the top 10% is not all better than aScore
        else:
            for i in range(len(result)):            
                if result[i] > aScore:
                    saved1.append(population[i])
        result_copy.sort()
##replace the bad chromosomes in the current generation with the good ones from the parent generation
        for i in range(len(population)):
            if saved0 == []:
                break
            if result[i] <= result_copy[len(saved0)-1]:
                population[i] = saved0.pop()
##save the good chromosome in the current generation as the good ones from parent 
        saved0 = saved1[:]
        result_copy.reverse()

##if finds a solution
        if result_copy[0] > endPoint:
            print("Hey, I found it")
            best_index = result.index(result_copy[0])
            print("the gene is:")
            print(population[best_index])
            print("It can run " + str(result_copy[0]) + " meters.")
            pop = []
            pop.append(population[best_index])
            fitness(pop,data)
            end = "yes"
    return pop
    
main()
